File: objectModel/CSharp/Microsoft.CommonDataModel.ObjectModel/Persistence/ModelJson/ExtensionHelper.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

using Microsoft.CommonDataModel.ObjectModel.Cdm;
using Microsoft.CommonDataModel.ObjectModel.Enums;
using Microsoft.CommonDataModel.ObjectModel.Persistence.ModelJson.types;
using Microsoft.CommonDataModel.ObjectModel.Utilities.Logging;
using Newtonsoft.Json.Linq;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace Microsoft.CommonDataModel.ObjectModel.Persistence.ModelJson
{
    static class ExtensionHelper
    {
        private static readonly string Tag = nameof(ExtensionHelper);

        /// <summary>
        /// Dictionary used to cache documents with trait definitions by file name.
        /// </summary>
        private static Dictionary<Tuple<CdmCorpusContext, string>, CdmDocumentDefinition> CachedDefDocs = new Dictionary<Tuple<CdmCorpusContext, string>, CdmDocumentDefinition>();

        /// <summary>
        /// Set of extensions that are officially supported and have their definitions in the extensions folder.
        /// </summary>
        private static readonly HashSet<string> SupportedExtensions = new HashSet<string>() { "pbi" };

        /// <summary>
        /// Constant used to store the prefix that is the mark of extension traits.
        /// </summary>
        public const string ExtensionTraitNamePrefix = "is.extension.";

        public const string ExtensionDocName = "custom.extension.cdm.json";

        /// <summary>
        /// Adds the list of documents with extensions schema definitions to the manifest.
        /// </summary>
        /// <param name="ctx">The context.</param>
        /// <param name="importDocs">The list of paths of documents with relevant schema definitions.</param>
        /// <param name="manifest">The manifest that needs to import the docs.</param>
        /// <returns></returns>
        internal static void AddImportDocsToManifest(CdmCorpusContext ctx, List<CdmImport> importDocs, CdmDocumentDefinition document)
        {
            foreach (var importDoc in importDocs)
            {
                if (!document.Imports.Any((CdmImport importPresent) => importPresent.CorpusPath == importDoc.CorpusPath))
                {
                    document.Imports.Add(importDoc);
                }
            }
        }

        /// <summary>
        /// For all the definitions (name, type) we have found for extensions, search in CDM Standard Schema for definition files.
        /// If we find a definition in a CDM Standard Schema file, we add that file to importsList.
        /// At the same time, the found definition is removed from extensionTraitDefList.
        /// When this function returns, extensionTraitDefList only contains definitions that are not present in any CDM Standard Schema file,
        /// and a list of CDM Standard Schema files with relevant definitions is returned.
        /// </summary>
        /// <param name="ctx"> The context</param>
        /// <param name="extensionTraitDefList"> The list of all definitions for all found extensions. Function modifies this list by removing definitions found in CDM Standard Schema files.</param>
        /// <param name="localExtensionTraitDefList"> The list of all definitions for all found extensions in local scope.</param>
        /// <returns> A list of CDM Standard Schema files to import.</returns>
        public static async Task<List<CdmImport>> StandardImportDetection(CdmCorpusContext ctx, List<CdmTraitDefinition> extensionTraitDefList, List<CdmTraitDefinition> localExtensionTraitDefList)
        {
            List<CdmImport> importsList = new List<CdmImport>();
            var hasCustomExtensionImport = false;

            // have to go from end to start because I might remove elements
            for (int traitIndex = localExtensionTraitDefList.Count - 1; traitIndex >= 0; traitIndex--)
            {
                CdmTraitDefinition extensionTraitDef = localExtensionTraitDefList[traitIndex];
                if (!TraitDefIsExtension(extensionTraitDef))
                {
                    Logger.Error(ctx, Tag, nameof(StandardImportDetection), extensionTraitDef.AtCorpusPath, CdmLogCode.ErrPersistModelJsonInvalidExtensionTrait, extensionTraitDef.TraitName, ExtensionTraitNamePrefix);
                    return null;
                }

                string[] extensionBreakdown = RemoveExtensionTraitNamePrefix(extensionTraitDef.TraitName).Split(':');
                if (extensionBreakdown.Length > 1)
                {
                    string extensionName = extensionBreakdown[0];

                    if (!SupportedExtensions.Contains(extensionName))
                    {
                        if (!hasCustomExtensionImport)
                        {
                            CdmImport importObject = ctx.Corpus.MakeObject<CdmImport>(CdmObjectType.Import);
                            importObject.CorpusPath = ExtensionDocName;
                            importsList.Add(importObject);
                            hasCustomExtensionImport = true;
                        }
                        continue;
                    }

                    string fileName = $"{extensionName}.extension.cdm.json";
                    string fileCorpusPath = $"cdm:/extensions/{fileName}";
                    CdmDocumentDefinition extensionDoc = await FetchDefDoc(ctx, fileName);

                    // If no document was found for that extensionName, the trait does not have a document with it's definition.
                    // Trait will be kept in extensionTraitDefList (a document with its definition will be created locally)
                    if (extensionDoc == null)
                    {
                        continue;
                    }

                    // There is a document with extensionName, now we search for the trait in the document.
                    // If we find it, we remove the trait from extensionTraitDefList and add the document to imports.
                    var matchingTrait = extensionDoc.Definitions.AllItems.Find(
                        (definition) => definition.ObjectType == CdmObjectType.TraitDef && definition.GetName() == extensionTraitDef.TraitName)
                        as CdmTraitDefinition;
                    if (matchingTrait != null)
                    {
                        List<CdmParameterDefinition> parameterList = matchingTrait.Parameters.AllItems;
                        if (
                            extensionTraitDef.Parameters.AllItems.TrueForAll(
                                (CdmParameterDefinition extensionParameter) =>
                                parameterList.Exists(
                                    (CdmParameterDefinition defParameter) => defParameter.Name == extensionParameter.Name)
                                )
                            )
                        {
                            extensionTraitDefList.Remove(extensionTraitDef);
                            if (!importsList.Exists((CdmImport importDoc) => importDoc.CorpusPath == fileCorpusPath))
                            {
                                CdmImport importObject = ctx.Corpus.MakeObject<CdmImport>(CdmObjectType.Import);
                                importObject.CorpusPath = fileCorpusPath;
                                importsList.Add(importObject);
                            }
                        }
                    }
                }
            }
            return importsList;
        }

        /// <summary>
        /// Processes extensions from an object which was obtained from a "model.json" file.
        /// From every extension found, it's value (name, value) is added to traitRefSet,
        /// and it's definition (name, type) is added to extensionTraitDefList.
        /// </summary>
        /// <param name="ctx"> The context </param>
        /// <param name="sourceObject"> The object obtained from "model.json" file.</param>
        /// <param name="traitRefSet"> The list of extensions found, in the form of (name & value).</param>
        /// <param name="extensionTraitDefList"> The list of definitions. For each extension, it's definition is added to this list (name & type).</param>
        /// <param name="localExtensionTraitDefList"> Same as extensionTraitDefList but limited to extensions inside one document.</param>
        public static void ProcessExtensionFromJson(
            CdmCorpusContext ctx,
            MetadataObject sourceObject,
            CdmTraitCollection traitRefSet,
            List<CdmTraitDefinition> extensionTraitDefList,
            List<CdmTraitDefinition> localExtensionTraitDefList = null)
        {
            var extensions = sourceObject.ExtensionFields;

            foreach (JProperty extensionAsJProperty in extensions.Children())
            {
                string traitName = AddExtensionTraitNamePrefix(extensionAsJProperty.Name);
                CdmTraitDefinition extensionTraitDef = extensionTraitDefList.Find((CdmTraitDefinition trait) => trait.TraitName == traitName);
                bool traitExists = extensionTraitDef != null;
                if (!traitExists)
                {
                    extensionTraitDef = ctx.Corpus.MakeObject<CdmTraitDefinition>(CdmObjectType.TraitDef, traitName);
                    extensionTraitDef.ExtendsTrait = ctx.Corpus.MakeObject<CdmTraitReference>(CdmObjectType.TraitRef, "is.extension", true);
                }

                CdmTraitReference extensionTraitRef = ctx.Corpus.MakeObject<CdmTraitReference>(CdmObjectType.TraitRef, traitName);
                var extensionValue = extensionAsJProperty.Value;
                var extensionType = extensionValue.Type;

                var shouldLookIntoProperties = extensionType == JTokenType.Object;

                if (shouldLookIntoProperties)
                {
                    var extensionValueAsJObject = extensionValue as JObject;
                    // extensionProperty.Name is here the equivalent of extensionProperty from TS project
                    foreach (JProperty extensionProperty in extensionValueAsJObject.Children())
                    {
                        // for every extensionProperty will have to save the (name, value) pair into extensionArgument,
                        // which will be saved in extensionTraitRef (the entity that will contain the data)
                        // (name, type) will be saved in extensionParameter,
                        // which will be saved in extensionTraitDef (the definition of the data, that can be saved in a schema file)
                        JToken extensionPropertyValue = extensionProperty.Value;

                        if (extensionPropertyValue == null)
                        {
                            continue;
                        }

                        CdmArgumentDefinition extensionArgument = ctx.Corpus.MakeObject<CdmArgumentDefinition>(
                            CdmObjectType.ArgumentDef,
                            extensionProperty.Name
                            );
                        CdmParameterDefinition extensionParameter = extensionTraitDef.Parameters.AllItems.Find(
                            (CdmParameterDefinition parameter) => parameter.Name == extensionProperty.Name);
                        bool parameterExists = extensionParameter != null;
                        if (!parameterExists)
                        {
                            extensionParameter = ctx.Corpus.MakeObject<CdmParameterDefinition>(CdmObjectType.ParameterDef, extensionProperty.Name);
                            extensionParameter.DataTypeRef = ctx.Corpus.MakeObject<CdmDataTypeReference>(CdmObjectType.DataTypeRef,
                                ConvertJTokenTypeToExpectedString(extensionPropertyValue.Type),
                                true);
                        }

                        if (extensionPropertyValue is JValue extensionPropertyValueAsJValue)
                        {
                            extensionArgument.Value = extensionPropertyValueAsJValue.Value;
                        }
                        else
                        {
                            extensionArgument.Value = extensionPropertyValue;
                        }

                        extensionTraitRef.Arguments.Add(extensionArgument);
                        if (!parameterExists)
                        {
                            extensionTraitDef.Parameters.Add(extensionParameter);
                        }
                    }
                }
                else
                {
                    CdmArgumentDefinition extensionArgument = ctx.Corpus.MakeObject<CdmArgumentDefinition>(CdmObjectType.ArgumentDef, traitName);
                    CdmParameterDefinition extensionParameter = extensionTraitDef.Parameters.AllItems.Find((CdmParameterDefinition parameter) => parameter.Name == traitName);
                    bool parameterExists = extensionParameter != null;
                    if (!parameterExists)
                    {
                        extensionParameter = ctx.Corpus.MakeObject<CdmParameterDefinition>(CdmObjectType.ParameterDef, traitName);
                        extensionParameter.DataTypeRef = ctx.Corpus.MakeObject<CdmDataTypeReference>(
                            CdmObjectType.DataTypeRef,
                            ConvertJTokenTypeToExpectedString(extensionType),
                            true);
                    }

                    if (extensionValue is JValue extensionValuePrimite)
                    {
                        extensionArgument.Value = extensionValuePrimite.Value;
                    }
                    else
                    {
                        extensionArgument.Value = extensionValue;
                    }

                    extensionTraitRef.Arguments.Add(extensionArgument);
                    if (!parameterExists)
                    {
                        extensionTraitDef.Parameters.Add(extensionParameter);
                    }
                }

                if (!traitExists)
                {
                    extensionTraitDefList.Add(extensionTraitDef);
                }

                if (localExtensionTraitDefList != null)
                {
                    localExtensionTraitDefList.Add(extensionTraitDef);
                }

                traitRefSet.Add(extensionTraitRef);
            }
        }

        /// <summary>
        /// Used as helper for converting a Manifest to a Model.
        /// Adds an extension stored in "Manifest" format to the data structure representing a <see cref="Model"/>
        /// </summary>
        /// <param name="extensionTraitRef">
        /// The data structure containing the extension in the format used by the Manifest
        /// </param>
        /// <param name="destination">
        /// The data structure used by <see cref="Model"/> where the data will be added to.
        /// There are multiple data structures that can have extensions, and any of these can be used here (assuming they are used by Model.Json data format)
        /// </param>
        public static void ProcessExtensionTraitToObject(CdmTraitReference extensionTraitRef, MetadataObject destination)
        {
            if (destination.ExtensionFields == null)
            {
                destination.ExtensionFields = new JObject();
            }

            string originalExtensionName = RemoveExtensionTraitNamePrefix(extensionTraitRef.NamedReference);
            JToken extensionValue;

            if (extensionTraitRef.Arguments.Count == 1 && extensionTraitRef.Arguments[0].Name == extensionTraitRef.NamedReference)
            {
                extensionValue = extensionTraitRef.Arguments[0].Value;
            }
            else
            {
                var extensionValueAsJObject = new JObject();
                foreach (CdmArgumentDefinition argument in extensionTraitRef.Arguments)
                {
                    var propertyName = argument.Name;
                    var propertyValue = argument.Value;
                    extensionValueAsJObject.Add(propertyName, propertyValue);
                }

                extensionValue = extensionValueAsJObject;
            }

            destination.ExtensionFields.Add(originalExtensionName, extensionValue);
        }

        /// <summary>
        /// Checks whether a trait name has the specific mark of an extension (begins with <see cref="ExtensionTraitNamePrefix"/>)
        /// </summary>
        /// <param name="traitName">The name of the trait to be checked.</param>
        /// <returns>Whether the trait is an extension.</returns>
        public static bool TraitNameHasExtensionMark(string traitName)
        {
            if (string.IsNullOrEmpty(traitName))
            {
                return false;
            }

            return traitName.StartsWith(ExtensionTraitNamePrefix);
        }

        /// <summary>
        /// Checks whether a <see cref="CdmTraitReferenceBase"/> is an extension.
        /// </summary>
        /// <param name="trait">The trait to be checked whether it is an extension.</param>
        /// <returns>Whether the trait is an extension.</returns>
        public static bool TraitRefIsExtension(CdmTraitReferenceBase trait)
        {
            return TraitNameHasExtensionMark(trait.NamedReference);
        }

        /// <summary>
        /// Checks whether a <see cref="CdmTraitDefinition"/> is an extension.
        /// </summary>
        /// <param name="trait">The trait to be checked whether it is an extension.</param>
        /// <return>Whether the trait is an extension.</return>
        public static bool TraitDefIsExtension(CdmTraitDefinition trait)
        {
            return TraitNameHasExtensionMark(trait.TraitName);
        }

        /// <summary>
        /// Tries to fetch the document with expected fileName.
        /// Caches results in <see cref="CachedDefDocs"/>.
        /// </summary>
        /// <param name="ctx">The context</param>
        /// <param name="fileName">The name of the file that needs to be retrieved.</param>
        /// <returns>The content of the definition file with the expected fileName, or null if no such file was found.</returns>
        private static async Task<CdmDocumentDefinition> FetchDefDoc(CdmCorpusContext ctx, string fileName)
        {
            // Since the CachedDefDocs is a static property and there might be multiple corpus running,
            // we need to make sure that each corpus will have its own cached def document.
            // This is achieved by adding the context as part of the key to the document.
            var key = Tuple.Create(ctx, fileName);
            if (CachedDefDocs.ContainsKey(key))
            {
                return CachedDefDocs[key];
            }

            string path = $"/extensions/{fileName}";
            var absPath = ctx.Corpus.Storage.CreateAbsoluteCorpusPath(path, ctx.Corpus.Storage.FetchRootFolder("cdm"));
            CdmObject document = await ctx.Corpus.FetchObjectAsync<CdmObject>(absPath);
            if (document != null)
            {
                CdmDocumentDefinition extensionDoc = document as CdmDocumentDefinition;

                // Needs to lock the dictionary since it is a static property and there might be multiple corpus running on the same environment.
                lock (CachedDefDocs)
                {
                    CachedDefDocs.Add(key, extensionDoc);
                }
                return extensionDoc;
            }
            return null;
        }

        /// <summary>
        /// Converts JTokenType to a string representing the type as expected in a Manifest.
        /// </summary>
        /// <param name="type">
        /// The type of the extension in the format used to deserialize Model.Json (JToken)
        /// </param>
        /// <returns>
        /// One of the allowed strings representing a type in a Manifest.
        /// </returns>
        private static string ConvertJTokenTypeToExpectedString(JTokenType type)
        {
            var stringType = type.ToString().ToLower();
            if (stringType.Equals("array"))
            {
                stringType = "object";
            }
            return stringType;
        }

        /// <summary>
        /// Context: To mark a trait as an extension, a prefix is added to the trait name.
        /// This function does the oposite; given a trait name with the extension prefix, it removes the prefix.
        /// </summary>
        /// <param name="traitName">The trait name with the extension prefix.</param>
        /// <returns>The trait name after the prefix was removed.</returns>
        private static string RemoveExtensionTraitNamePrefix(string traitName)
        {
            return traitName.Substring(ExtensionTraitNamePrefix.Length);
        }

        /// <summary>
        /// Adds a prefix to a trait name, marking it as an extension (<see cref="ExtensionTraitNamePrefix"/>)
        /// </summary>
        /// <param name="traitName">The name of the trait to be marked as extension.</param>
        /// <returns>The trait name with the prefix that marks an extension.</returns>
        private static string AddExtensionTraitNamePrefix(string traitName)
        {
            return $"{ExtensionTraitNamePrefix}{traitName}";
        }
    }
}

File: objectModel/Python/cdm/persistence/modeljson/extension_helper.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See License.txt in the project root for license information.

from collections import OrderedDict
from typing import List, Dict, Optional, Tuple, TYPE_CHECKING

from cdm.enums import CdmObjectType, CdmLogCode
from cdm.utilities import JObject, logger

if TYPE_CHECKING:
    from cdm.objectmodel import CdmCorpusContext, CdmDocumentDefinition, CdmImport, CdmTraitCollection, CdmTraitDefinition

cached_def_docs = {}  # type: Dict[Tuple[CdmCorpusContext, str], CdmDocumentDefinition]

convert_type_to_expected_string = {
    str: 'string',
    int: 'number',
    dict: 'object',
    OrderedDict: 'object',
    bool: 'boolean',
    list: 'object',
    type(None): 'object'
}

_TAG = 'ExtensionHelper'

EXTENSION_TRAIT_NAME_PREFIX = 'is.extension.'
EXTENSION_DOC_NAME = 'custom.extension.cdm.json'

# Set of extensions that are officially supported and have their definitions in the extensions folder.
SUPPORTED_EXTENSIONS = {"pbi"}


def add_import_docs_to_manifest(ctx: 'CdmCorpusContext', import_docs: List['CdmImport'], document: 'CdmManifestDefinition'):
    for import_doc in import_docs:
        if not list(filter(lambda import_present: import_present.corpus_path == import_doc.corpus_path, document.imports)):
            document.imports.append(import_doc)


async def fetch_def_doc(ctx: 'CdmCorpusContext', file_name: str) -> None:
    # Since the CachedDefDocs is a static property and there might be multiple corpus running,
    # we need to make sure that each corpus will have its own cached def document.
    # This is achieved by adding the context as part of the key to the document.
    key = (ctx, file_name)
    if key in cached_def_docs:
        return cached_def_docs[key]

    path = '/extensions/{}'.format(file_name)
    document = await ctx.corpus.fetch_object_async(path, ctx.corpus.storage.fetch_root_folder('cdm'))

    if document is not None:
        cached_def_docs[key] = document

    return document


async def standard_import_detection(ctx: 'CdmCorpusContext', extension_trait_def_list: List['CdmTraitDefinition'],
                                    local_extension_trait_def_list: List['CdmTraitDefinition']) -> Optional[List['CdmImport']]:
    imports_list = []
    trait_index = len(local_extension_trait_def_list) - 1
    has_custom_extension_import = False

    while trait_index >= 0:
        extension_trait_def = local_extension_trait_def_list[trait_index]

        if not extension_trait_def.trait_name or not extension_trait_def.trait_name.startswith(EXTENSION_TRAIT_NAME_PREFIX):
            logger.error(ctx, _TAG, standard_import_detection.__name__, extension_trait_def.at_corpus_path, CdmLogCode.ERR_PERSIST_INVALID_EXTENSION_TRAIT, extension_trait_def.trait_name, EXTENSION_TRAIT_NAME_PREFIX)
            return None

        extension_breakdown = extension_trait_def.trait_name[len(EXTENSION_TRAIT_NAME_PREFIX):].split(':')

        if len(extension_breakdown) > 1:
            extension_name = extension_breakdown[0]

            if extension_name not in SUPPORTED_EXTENSIONS:
                if not has_custom_extension_import:
                    import_object = ctx.corpus.make_object(CdmObjectType.IMPORT)
                    import_object.corpus_path = EXTENSION_DOC_NAME
                    imports_list.append(import_object)
                    has_custom_extension_import = True
                trait_index -= 1
                continue

            file_name = '{}.extension.cdm.json'.format(extension_name)
            file_corpus_path = 'cdm:/extensions/{}'.format(file_name)
            extension_doc = await fetch_def_doc(ctx, file_name)

            # if no document was found for that extensionName, the trait does not have a document with it's definition.
            # trait will be kept in extensionTraitDefList (a document with its definition will be created locally)
            if not extension_doc:
                trait_index -= 1
                continue

            # there is a document with extensionName, now we search for the trait in the document.
            # if we find it, we remove the trait from extensionTraitDefList and add the document to imports.
            matching_traits = [
                definition
                for definition in extension_doc.definitions
                if definition.object_type == CdmObjectType.TRAIT_DEF and definition.get_name() == extension_trait_def.trait_name
            ]

            if matching_traits:
                parameter_list = matching_traits[0].parameters

                if all(any(def_parameter.name == extension_parameter.name for def_parameter in parameter_list)
                       for extension_parameter in extension_trait_def.parameters):
                    extension_trait_def_list.remove(extension_trait_def)

                    if not any(import_doc.corpus_path == file_corpus_path for import_doc in imports_list):
                        import_object = ctx.corpus.make_object(CdmObjectType.IMPORT)
                        import_object.corpus_path = file_corpus_path
                        imports_list.append(import_object)

        trait_index -= 1

    return imports_list


def process_extension_from_json(ctx: 'CdmCorpusContext', extensions: object, trait_ref_set: 'CdmTraitCollection',
                                extension_trait_def_list: List['CdmTraitDefinition'],
                                local_extension_trait_def_list: Optional[List['CdmTraitDefinition']] = None) -> None:
    extension_keys = [extension for extension in extensions.keys() if extension.find(':') != -1]

    for extension_key in extension_keys:
        trait_name = 'is.extension.{}'.format(extension_key)
        extension_trait_defs = [trait for trait in extension_trait_def_list if trait.trait_name == trait_name]
        extension_trait_def = None
        trait_exists = bool(extension_trait_defs)

        if trait_exists:
            extension_trait_def = extension_trait_defs[0]
        else:
            extension_trait_def = ctx.corpus.make_object(CdmObjectType.TRAIT_DEF, trait_name)
            extension_trait_def.extends_trait = ctx.corpus.make_object(CdmObjectType.TRAIT_REF, 'is.extension', True)

        extension_trait_ref = ctx.corpus.make_object(CdmObjectType.TRAIT_REF, trait_name)
        extension_value = extensions[extension_key]
        is_array = isinstance(extension_value, list)

        if extension_value and isinstance(extension_value, dict) and not is_array:
            extension_properties = [extension for extension in extension_value.keys() if not extension.startswith('_')]
            for extension_property in extension_properties:
                extension_property_value = extension_value[extension_property]
                extension_argument = ctx.corpus.make_object(CdmObjectType.ARGUMENT_DEF, extension_property)
                extension_parameters = [parameter for parameter in extension_trait_def.parameters if parameter.name == extension_property]
                extension_parameter = None
                parameter_exists = bool(extension_parameters)

                if parameter_exists:
                    extension_parameter = extension_parameters[0]
                else:
                    extension_parameter = ctx.corpus.make_object(CdmObjectType.PARAMETER_DEF, extension_property)
                    extension_parameter.data_type_ref = ctx.corpus.make_object(
                        CdmObjectType.DATA_TYPE_REF,
                        convert_type_to_expected_string[type(extension_property_value)],
                        True)

                extension_argument.value = extension_property_value
                extension_trait_ref.arguments.append(extension_argument)

                if not parameter_exists:
                    extension_trait_def.parameters.append(extension_parameter)
        else:
            extension_argument = ctx.corpus.make_object(CdmObjectType.ARGUMENT_DEF, trait_name)
            extension_parameters = [
                parameter for parameter in extension_trait_def.parameters if parameter.name == trait_name
            ]
            extension_parameter = None
            parameter_exists = bool(extension_parameters)

            if parameter_exists:
                extension_parameter = extension_parameters[0]
            else:
                extension_parameter = ctx.corpus.make_object(CdmObjectType.PARAMETER_DEF, trait_name)
                extension_parameter.data_type_ref = ctx.corpus.make_object(
                    CdmObjectType.DATA_TYPE_REF,
                    convert_type_to_expected_string[type(extension_value)],
                    True)

            extension_argument.value = extension_value
            extension_trait_ref.arguments.append(extension_argument)

            if not parameter_exists:
                extension_trait_def.parameters.append(extension_parameter)

        if not trait_exists:
            extension_trait_def_list.append(extension_trait_def)

        if local_extension_trait_def_list is not None:
            local_extension_trait_def_list.append(extension_trait_def)

        trait_ref_set.append(extension_trait_ref)


def process_extension_trait_to_object(extension_trait_ref: 'CdmTraitReference', add_property_to: 'JObject'):
    # TODO: Another magic number
    original_prop_name = extension_trait_ref.named_reference[13:]
    if len(extension_trait_ref.arguments) == 1 and \
            extension_trait_ref.arguments[0].name == extension_trait_ref.named_reference:

        add_property_to[original_prop_name] = extension_trait_ref.arguments[0].value
        return

    value_object = {}

    for argument in extension_trait_ref.arguments:
        value_object[argument.name] = argument.value

    add_property_to[original_prop_name] = value_object
